rts/src/main/java/eta/runtime/io/MemoryManager.java

```diff
@@ -237,4 +242,161 @@ public static long loadStringsUTF8(String[] ss) throws UnsupportedEncodingExcept
+    }
+    
+    public static ByteBuffer copyByteBuffer( ByteBuffer src, ByteBuffer dest, int n) {
+        return copyByteBuffer(src, 0, dest, 0, n);
+    }
+
+    public static ByteBuffer copyByteBufferDups( ByteBuffer src, ByteBuffer dest, int n) {
+        return copyByteBufferDups(src, 0, dest, 0, n);
+    }
+    
+    public static ByteBuffer copyByteBuffer( ByteBuffer src, int srcOffset
+                                           , ByteBuffer dest, int destOffset
+                                           , int n) {
+        if (src == dest)
+            return dest;
+        src.position(src.position() + srcOffset);
+        src.limit(src.position() + n);
+        dest.position(dest.position() + destOffset);
+        dest.put(src);
+        return dest;
+    }
+
+    public static ByteBuffer copyByteBufferDups( ByteBuffer src, int srcOffset
+                                               , ByteBuffer dest, int destOffset
+                                               , int n) {
+        src = src.duplicate();
+        dest = dest.duplicate();
+        return copyByteBuffer(src, srcOffset, dest, destOffset, n);
+    }
+    
+    public static ByteBuffer copy( long srcAddress, int srcOffset
+                                 , ByteBuffer dest, int destOffset
+                                 , int n) {
+        ByteBuffer src = getBoundedBuffer(srcAddress);
+        return copyByteBuffer(src,srcOffset,dest,destOffset,n);
+    }
+
+    public static ByteBuffer copy( ByteBuffer src, int srcOffset
+                                 , long destAddress, int destOffset
+                                 , int n) {
+        ByteBuffer dest = getBoundedBuffer(destAddress);
+        return copyByteBuffer(src,srcOffset,dest,destOffset,n);
+    }
+
+    
+    public static ByteBuffer copy( long srcAddress, int srcOffset
+                                 , long destAddress, int destOffset
+                                 , int size) {
+        ByteBuffer src  = getBoundedBuffer(srcAddress);
+        ByteBuffer dest = getBoundedBuffer(destAddress);
+        return copyByteBuffer(src,srcOffset,dest,destOffset,size);
+    }
+
+    public static ByteBuffer copy(long srcAddress, long destAddress, int size) {
+        return copy(srcAddress, 0, destAddress, 0, size);
+    }
+
+    public static ByteBuffer set(long address, int val, int size) {
+        ByteBuffer buffer = getBoundedBuffer(address);
+        while (size-- != 0) {
+            buffer.put((byte) val);
+        }
+        return buffer;
+    }
+       
+    public static ByteBuffer set(long address, byte[] bytes) {
+        ByteBuffer buf = getBoundedBuffer(address);
+        buf.put(bytes);
+        return buf;
+    }
+
+    public static long allocateAndSet(byte[] bytes) {
+        long address = allocateBuffer(bytes.length,true);
+        set(address,bytes);
+        return address;
+    }
+    
+    public static ByteBuffer move(long srcAddress, long destAddress, int size) {
```

Comment:
Ok, I see you've put src and dest, consistent with the ordering elsewhere. I think when calling this in `c_memmove`, you swapped the arguments by mistake.

Reply:
yeah, thanks for catching it